Add type-safe, const-safe loops to c10

r-barnes opened the pull request:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46414

For loops are often written with mismatched data types which causes silent type and sign coercion in the absence of integer conversion warnings. Getting around this in templated code requires convoluted patterns such as
```
for(auto i=decltype(var){0};i<var;i++)
```
(the above pattern appears ~31 times in the code base) with this diff we can instead write
```
for(const auto i = c10::ranges::indices(var))
```
Note that this loop is type-safe and const-safe. (Type-unsafe loops occur an unknown number of times do to the lack of conversion warnings on compilation.)

The functions introduced here (`c10::ranges::ints` and `c10::ranges::indices`) allow for type-safety and const-ness within for loops, which prevents the accidental truncation or modification of integers and other types, improving code safety. Further, they use the same naming convention as the ranges-v3 library and offer an easy migration path to C++20 functionality.

Test Plan:
```
buck run //caffe2/c10:c10_test_0
```

Differential Revision: D24334732

ezyang commented:
I'm going to let @smessmer make the call on this one. This seems to be a Boost-ism, not part of upcoming C++ standard library.

r-barnes commented:
That's correct, @ezyang, C++20 or prior doesn't, as far as I can tell, have a similar functionality. I'll edit the commit message to include:
> `irange` allows for type-safety and const-ness within for loops, which prevents the accidental truncation of integers and other types, improving code safety.

r-barnes commented:
@ezyang or @smessmer : It seems like review on this has stalled, so I've simplified the diff. I've also expanded my comments on the rationale.

r-barnes commented:
#50484 , #50485 , #50482 , #50486 all fix compiler warnings and potential integer overflow/wrap issues that this PR would prevent.

ngimel commented on c10/util/irange.h:
```diff
@@ -0,0 +1,72 @@
+/// Creates an integer range for the half-open interval [begin, end)
+template <typename T>
+integer_range<T> irange(T begin, T end) {
+    TORCH_CHECK(begin <= end);
```
This is adding a TORCH_CHECK and might affect performance of hot paths

ngimel commented on c10/util/irange.h:
```diff
@@ -0,0 +1,72 @@
+
+/// Creates an integer range for the half-open interval [0, end)
+template <typename T>
+integer_range<T> irange(T end) {
```
What would happen if `end` is negative? Can you add a test for this?

ngimel reviewed (commented):
I'm not sure how to make it performance-neutral compared to (also type-safe) loops that are written with decltype. TORCH_CHECK is known to prevent inlining, and probably it would have to be added to an upper-limit variant too.

r-barnes committed:
Move irange to c10 (#46414)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46414

For loops are often written with mismatched data types which causes silent type and sign coercion in the absence of integer conversion warnings. Getting around this in templated code requires convoluted patterns such as
```
for(auto i=decltype(var){0};i<var;i++)
```
with this diff we can instead write
```
for(const auto i = c10::irange(var))
```
Note that this loop is type-safe and const-safe.

The function introduced here (`c10::irange`) allows for type-safety and const-ness within for loops, which prevents the accidental truncation or modification of integers and other types, improving code safety.

Test Plan:
```
buck test //caffe2/c10:c10_test_0
```

Differential Revision: D24334732

fbshipit-source-id: 595d03de995288da83d0e4325e269511a7f122f7

ngimel commented:
For the record, here's godbolt for irange compared to regular for loop, there's an extra cmovsq: https://godbolt.org/z/3o1968
